src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicBufferHandle.cs

```diff
@@ -0,0 +1,29 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic.Internal
+{
+    internal sealed unsafe class SafeMsQuicBufferHandle : SafeHandle
+    {
+        public int Count;
+
+        public override bool IsInvalid => handle == IntPtr.Zero;
+
+        public SafeMsQuicBufferHandle(int count)
+            : base(IntPtr.Zero, ownsHandle: true)
+        {
+            IntPtr buffer = Marshal.AllocHGlobal(sizeof(QuicBuffer) * count);
```

Comment:
We shouldn't need to do this? What pattern is causing this to be required?

Reply:
Do you mean `AllocHGlobal` or wrapping it in `SafeMsQuicBufferHandle`?

Reply:
Both.

Reply:
I agree that the SafeHandle should be unnecessary here. The code has to be careful about managing the lifetimes of the async buffers. SafeHandle does not make it easier.

It is an interesting question whether it is better to use unmanaged memory or POH-allocated memory here.

Reply:
@nibanks

Reply:
@geoffkizer what exactly is the question/problem here?

Reply:
Winsock does not require the buffer array to be pinned until the operation completes. (Buffers themselves must be pinned, of course.)

We are wondering why msquic is different than Winsock in this regard.

Reply:
The difference may be that the buffer is not consumed by the time the call finishes (like it is for Ssl) and need to wait for some event in future. 
We keep pinning for each chunk of data we send. The QuickBuffer is only used to pass data to native code so the AllocHGlobal look like best choice as we don't need to touch of every send. So it is not necessary I just thought it is better option. As far as the SafeBuffer or not. I did not use before but as @jkotas  mentioned we need to be careful. Particular when the object is disposed when the p/invioke is done but squick is still using the buffer.

Reply:
It is more efficient this way. Otherwise, MsQuic would have to dynamically allocate memory for each send ourselves. I'm not positive why WinSock is different, but suspect that's because they have to already dynamically allocate stuff for the IOCTL call so it's not an extra overhead. Since it's all in proc for MsQuic, we don't need to do that.

Reply:
Here's the equivalent struct in our libuv transport implementation https://github.com/dotnet/aspnetcore/blob/0011eff5648bc555e1a48711c7c24aea34afa785/src/Servers/Kestrel/Transport.Libuv/src/Internal/Networking/UvWriteReq.cs#L15. It doesn't use native memory, it pins the array of buffers for the lifetime of the operation and unpins afterwards. What makes managing the lifetime difficult here?

Reply:
For example you have to be careful about situation when somebody disposes the object while the operation is still in flight.

Your libuv example is doing explicit lifetime management for GCHandles that wrap managed memory. Instead of doing explicit lifetime management for GCHandles, you can equally well do the explicit lifetime management for unmanaged memory and avoid an extra indirection, etc. GCHandle is in fact a piece of unmanaged memory, allocate using a special purpose unmanaged memory allocator.

Reply:
Here is another example from SslStream. 
https://github.com/dotnet/runtime/blob/7ea26e68f9856df48fb6f48666fc643bbcbefbc6/src/libraries/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs#L262-L268

Because we need to only live through the one native call, we can allocate the buffer array on stack. For msquic we need to wait until `HandleEventSendComplete()` is invoked. So we need to keep it alive past the sending function.

Anyway, we could use the pinning if we want to but as I mentioned this looked simpler and less work on each call.

Reply:
> For example you have to be careful about situation when somebody disposes the object while the operation is still in flight.

Yep, I'm aware and we also go back and forth on how useful SafeHandles are in certain places, but I'm surprised SafeHandle + native buffer is the solution here. Are we pooling these things?

Reply:
there is no pool at the moment. The main goal was to stabilize the test runs and avoid corruption and crashes and to unblock @ManickaP  and @CarnaViire. I did 700+ runs and I did not see any failure. I think it would be better to refactor and/or optimize once we have solid CI in place. 

I think we will need more work toward  #5262 and #32142. The SafeHandle primarily helps with p/invokes. For writes, we will need to preserve the memory longer unless we change the handout model.